### PR TITLE
Implement the ForceDeleteAdminRequest

### PR DESCRIPTION
**Pull request by JingQianCloud**
Implement the ForceDeleteAdminRequest.
Added config to enable/disable ForceDelete request.

### CODE REVIEW

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java [hunk 1]
```diff
@@ -536,6 +548,56 @@ private AdminResponse handleBlobIndexRequest(DataInputStream requestStream, Admi
...
+      // Get the size of the delete record
+      MessageFormatInputStream stream =
+          new DeleteMessageFormatInputStream(storeKey, accountId, containerId, operationTimeMs, lifeVersion);
+      // Generate the MessageInfo
+      MessageInfo info = new MessageInfo(storeKey, stream.getSize(), accountId, containerId, operationTimeMs, lifeVersion);
+
+      store.forceDelete(Collections.singletonList(info));
+      return new AdminResponse(correlationId, clientId, ServerErrorCode.No_Error);
+    } catch (StoreException e) {
+      logger.error("ForceDeleteAdminRequest Failed to find all message infos for given key {}", blobId, e);
```
Comment (justinlin-linkedin) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java:591]: log message seems to be a bit off.

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java [hunk 2]
```diff
@@ -536,6 +548,56 @@ private AdminResponse handleBlobIndexRequest(DataInputStream requestStream, Admi
         errorMessage.getBytes());
   }
 
+  /**
+   * Handles {@link ForceDeleteAdminRequest}.
+   * @param requestStream the serialized bytes of the request.
+   * @param adminRequest the {@link AdminRequest} received.
+   * @return the {@link AdminResponse} to the request.
+   */
+  private AdminResponse handleForceDeleteRequest(DataInputStream requestStream, AdminRequest adminRequest) {
```
Comment (justinlin-linkedin) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java:557]: let's add a configuration to turn on and off this admin request, we want to turn this by default in prod.

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java [hunk 3]
```diff
@@ -479,6 +484,72 @@ static void endToEndTest(Port targetPort, String routerDatacenter, MockCluster c
...
+      // Use BlobIndexAdminRequest to verify we have one message info.
+      blobIndexAdminRequest = new BlobIndexAdminRequest(blobId6,
+          new AdminRequest(AdminRequestOrResponseType.BlobIndex, partitionIds.get(0), 1, "clientid2"));
+      stream = channel.sendAndReceive(blobIndexAdminRequest).getInputStream();
+      adminResponseWithContent = AdminResponseWithContent.readFrom(stream);
+      releaseNettyBufUnderneathStream(stream);
+      assertEquals(new String(adminResponseWithContent.getContent()), ServerErrorCode.No_Error,
+          adminResponseWithContent.getError());
+      jsonBytes = adminResponseWithContent.getContent();
+      objectMapper = new ObjectMapper();
```
Comment (justinlin-linkedin) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:508]: we don't have to create objectMapper again.

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java [hunk 4]
```diff
@@ -479,6 +484,72 @@ static void endToEndTest(Port targetPort, String routerDatacenter, MockCluster c
...
+      StoreKeyJacksonConfig.setupObjectMapper(objectMapper, new BlobIdFactory(clusterMap));
```
Comment (justinlin-linkedin) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:509]: same here

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java [hunk 5]
```diff
@@ -479,6 +484,72 @@ static void endToEndTest(Port targetPort, String routerDatacenter, MockCluster c
...
+      releaseNettyBufUnderneathStream(stream);
+      assertEquals(ServerErrorCode.Blob_Already_Exists, adminResponse.getError());
+
+      // Do this test for a fake partition
+      forceDeleteAdminRequest = new ForceDeleteAdminRequest(fakeBlobId, lifeVersion,
+          new AdminRequest(AdminRequestOrResponseType.ForceDelete, fakePartition, 1, "clientid2"));
+      stream = channel.sendAndReceive(forceDeleteAdminRequest).getInputStream();
+      adminResponse = AdminResponse.readFrom(stream);
+      releaseNettyBufUnderneathStream(stream);
+      assertEquals(ServerErrorCode.Bad_Request, adminResponse.getError());
```
Comment (justinlin-linkedin) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:540]: shouldn't this be replica_unavailable error?

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java (hunk 4 above)
Comment (JingQianCloud) [reply ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:509]: For the following two lines,
objectMapper = new ObjectMapper();
StoreKeyJacksonConfig.setupObjectMapper(objectMapper, new BlobIdFactory(clusterMap));

I thought I didn't need these two lines either, but I have to add them back, missing one will cause the exception somehow. 

Cannot construct instance of `com.github.ambry.store.StoreKey` (no Creators, like default constructor, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information
```suggestion
      StoreKeyJacksonConfig.setupObjectMapper(objectMapper, new BlobIdFactory(clusterMap));
```

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java (hunk 2 above)
Comment (JingQianCloud) [reply ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java:557]: Added.

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java (hunk 4 above)
Comment (JingQianCloud) [reply ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:509]: Looks like it's related to the following line.

> https://jarvis.corp.linkedin.com/codesearch/result/?name=ServerTestUtil.java&path=ambry%2Fambry-server%2Fsrc%2Fintegration-test%2Fjava%2Fcom%2Fgithub%2Fambry%2Fserver&reponame=linkedin%2Fambry#458
 
After I removed the above line, I don't need reinitialize it any more.

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java (hunk 3 above)
Comment (JingQianCloud) [reply ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:508]: It was due to some extra new ObjectManager as the other comment.

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java [hunk 6]
```diff
@@ -536,6 +548,62 @@ private AdminResponse handleBlobIndexRequest(DataInputStream requestStream, Admi
...
+      PartitionId partitionId = blobId.getPartition();
+      Store store = storeManager.getStore(partitionId);
+      if (store == null) {
+        logger.error("ForceDeleteAdminRequest BlobId {}'s partition[{}] doesn't exist or stopped in this host", blobId, partitionId.getId());
+        return new AdminResponse(correlationId, clientId, ServerErrorCode.Replica_Unavailable);
+      }
+      // Get the size of the delete record
+      MessageFormatInputStream stream =
+          new DeleteMessageFormatInputStream(storeKey, accountId, containerId, operationTimeMs, lifeVersion);
+      // Generate the MessageInfo
```
Comment (justinlin-linkedin) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java:591]: we don't need to create a stream here, you can just pass a message info to the forceDelete method. We will recalculate the size in the forceDelete.

Comment (JingQianCloud) [reply ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java]: Removed. Thanks,

**Review by justinlin-linkedin — APPROVED** [review]

**Review by SophieGuo410 — APPROVED** [review]

**Pull request merged**